From: Hannes Reinecke <hare@suse.de>
Date: Wed, 7 Mar 2018 12:52:52 +0100
Subject: [PATCH] nvme_rdma: clear NVME_RDMA_Q_LIVE bit if reconnect fails
References: bsc#1083770
Patch-Mainline: Never, SLE12-SP3 specific

When starting reconnection we correctly unset the NVME_RDMA_Q_LIVE
bit, and set if once the reconnect is attempted.
However, we never clear that bit again if the reconnect attempt
failed.

Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 drivers/nvme/host/rdma.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/nvme/host/rdma.c b/drivers/nvme/host/rdma.c
index f58101f5a7b3..50b45557a085 100644
--- a/drivers/nvme/host/rdma.c
+++ b/drivers/nvme/host/rdma.c
@@ -644,6 +644,8 @@ static int nvme_rdma_connect_io_queues(struct nvme_rdma_ctrl *ctrl)
 	return 0;
 
 out_free_queues:
+	for (i = 1; i < ctrl->queue_count; i++)
+		clear_bit(NVME_RDMA_Q_LIVE, &ctrl->queues[i].flags);
 	nvme_rdma_free_io_queues(ctrl);
 	return ret;
 }
@@ -800,6 +802,7 @@ static void nvme_rdma_reconnect_ctrl_work(struct work_struct *work)
 requeue:
 	dev_info(ctrl->ctrl.device, "Failed reconnect attempt %d\n",
 			ctrl->ctrl.opts->nr_reconnects);
+	clear_bit(NVME_RDMA_Q_LIVE, &ctrl->queues[0].flags);
 	nvme_rdma_reconnect_or_remove(ctrl);
 }
 
-- 
2.12.3

